### PR TITLE
Fix TypeError when switching snippet template with block fields

### LINKED ISSUE #8697: Bug: TypeError: g.map is not a function when switching snippet template with block fields
**Sulu version:** 3.0.4

**Description:**
When switching a snippet template in the admin panel (e.g. from "footer" to "menu"), 
the admin crashes with `TypeError: g.map is not a function` if both templates contain block fields.

**Steps to reproduce:**
1. Create two snippet templates with block fields (e.g. footer.xml and menu.xml)
2. Go to Snippets → Add
3. Select the first template (e.g. footer)
4. Switch to the second template (e.g. menu)
5. Admin crashes

**Expected behavior:**
Template switches cleanly, block fields reset to empty state.

**Actual behavior:**
```
TypeError: g.map is not a function at SortableBlockList.js:154
RangeError: Invalid array length at BlockCollection.js:188
TypeError: n.filter is not a function at BlockCollection.js:130
```

**Root cause:**
When switching templates, Sulu passes block data from the old template to the new 
`BlockCollection` component. The data arrives as a plain object `{}` instead of an 
array `[]`. The component calls `.map()`, `.filter()` and `.length` directly on the 
object without checking its type first, causing a crash.

**Fix:**
Add `Array.isArray()` checks before array operations in `BlockCollection.js` and 
`SortableBlockList.js`:

- `BlockCollection.js`: add `!Array.isArray(value)` guard in `fillArrays`, 
`ensureBlockIds` and the reaction
- `SortableBlockList.js`: add `Array.isArray(value)` check before `value.map()` #

### PR DESCRIPTION
**Pull request by Amoifr**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #8697
| Related issues/PRs | 
| License | MIT
| Documentation PR | 

#### What's in this PR?

This PR adds `Array.isArray()` checks to prevent crashes when switching snippet templates that contain block fields.

**Root cause:** When switching templates, block data from the old template arrives as an object `{}` instead of an array `[]`. The component calls `.map()`, `.filter()`, `.length` directly on this value without type checking.

**Solution:** Added an `ensureArray()` helper function and used it throughout `BlockCollection.js` before any array operations.

**Changes:**
- `BlockCollection.js`: Added `ensureArray()` helper and applied it in:
  - Constructor reaction
  - `componentDidUpdate`
  - `ensureBlockIds`
  - `fillArrays`
  - `handleAddBlock`, `handlePasteBlocks`, `removeBlocks`, `duplicateBlocks`, `copyBlocks`, `cutBlocks`
  - Computed getters (`hasMaximumReached`, `hasMinimumReached`)
  - Render methods (`renderAddButton`, `renderBlockToolbar`, `render`)
- `SortableBlockList.js`: Added `Array.isArray()` check before `value.map()`
- Added 3 tests for handling object, null, and undefined values

#### Why?

The admin panel crashes with `TypeError: g.map is not a function` when switching between snippet templates that both contain block fields.

#### ⚠️ Note

We may have jumped too quickly into this PR. The scope of changes might be broader than what the original issue described. Only the maintainers can properly assess whether this approach is the right one and if the fix covers all edge cases correctly.

The reporter suggested adding checks in `fillArrays`, `ensureBlockIds`, and `SortableBlockList.js`, but we found that many other methods also needed protection. This comprehensive approach might be over-engineered, or it might be exactly what's needed - we defer to the maintainers' judgment.

#### To Do

- [ ] ~Create a documentation PR~ (no documentation changes needed)
- [x] ~Add breaking changes to UPGRADE.md~ (no breaking changes)

### CODE REVIEW

**Comment by Amoifr** [comment]
Hey everyone! 👋

I just pushed an update that includes additional tests inspired by the XML files shared in #8697.

I wanted to take a moment to warmly thank:

- **@alexander-schranz** for the thorough review and for pointing out the need for concrete test cases. Your guidance helped shape this fix into something much more robust! 🙏
- **@Leanid554** for taking the time to provide the `menu.xml` and `footer.xml` files. These real-world examples with deeply nested blocks (3 levels deep!) were invaluable for understanding and testing the edge cases. 

The new tests now cover:
- Nested blocks becoming objects when switching templates (menu.xml scenario)
- Deep 3-level nesting (pages → subpages → subpages2)
- Multiple nested block types (pages + socials from footer.xml)
- Full lifecycle transitions: array → object → array

All 79 tests pass! ✅

Thanks again for your collaboration – open source is truly a team effort! 🎉

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js [hunk 1]
```diff
@@ -159,89 +178,104 @@ class BlockCollection<T: string, U: {_id?: string, type: T, ...}> extends React.
...
-                const newBlockCount = minOccurs - value.length;
+        this.isFillingArrays = true;
 
-                // Create blocks and generate IDs if needed
-                let newBlocks;
-                if (generateBlockIds) {
-                    try {
-                        const blockIds = await generateBlockIds(newBlockCount);
+        try {
+            const blockIds = await generateBlockIds(newBlockCount);
```
Comment (Prokyonn) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js]: Please keep the `fillArrays` rewrite out of this fix: the crash does not need it, and a rejected `generateBlockIds` is no longer caught here, so it becomes an unhandled promise rejection.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/BlockCollection.test.js [hunk 2]
```diff
@@ -1829,3 +1829,174 @@ test('Should not generate IDs when generateBlockIds is not provided', async() =>
...
+
+test('Should handle undefined value gracefully', () => {
+    // $FlowFixMe - intentionally passing wrong type to test error handling
+    renderBlockCollection({value: undefined});
+
+    // Should render without crashing, with no blocks visible
+    expect(getBlocks()).toHaveLength(0);
+});
+
+test('Should handle blocks with nested object properties when switching templates (menu.xml scenario)', () => {
```
Comment (Prokyonn) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/BlockCollection.test.js]: This test and the next two also pass without the fix, as does the `undefined` one above: `renderBlockContent` only stringifies the nested value, so no nested `BlockCollection` ever receives it. Please drop them.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/fixtures/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# BlockCollection Test Fixtures
```
Comment (Prokyonn) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/fixtures/README.md]: Please remove the `fixtures/` directory: no test loads these files.

**Review by Prokyonn — CHANGES_REQUESTED** [review]
Could you skip the remote request in `ResourceFormStore.loadAndMergeRemoteData` when there is no id, instead of guarding `BlockCollection`? On an add form the template switch loads the snippet list, and its `pages: 1` lands in the block named `pages`; the guards stop the crash but still save that `1`.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js (hunk 1 above)
Comment (Amoifr) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js]: Dropped: `BlockCollection.js` is untouched now, the fix moved to `ResourceFormStore` as you suggested.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/BlockCollection.test.js (hunk 2 above)
Comment (Amoifr) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/BlockCollection.test.js]: Right, they couldn't fail. All `BlockCollection` tests are gone; the new test is in `ResourceFormStore.test.js` and fails without the fix.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/fixtures/README.md (hunk 3 above)
Comment (Amoifr) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/tests/fixtures/README.md]: Removed, along with the rest of the `BlockCollection` changes.

**Comment by Amoifr** [comment]
@Prokyonn thanks, that's the real cause: without an id, `requestRemoteData()` sends `id: undefined` and gets the list response, whose `pages: 1` lands in the block. `loadAndMergeRemoteData()` now skips the request when there is no id, like `ResourceStore.load()` does, and still merges with empty remote data so the values that fit the new template are kept. The new test fails without it: the request goes out and `pages` receives `1`.

Since `2.6` has the same code, I moved the PR there and rebuilt it as a single commit; the `BlockCollection` guards, the extra tests and the fixtures are all gone.
